py-polars/polars/internals/functions.py

```diff
@@ -371,3 +371,98 @@ def cut(
+) -> list[pli.DataFrame]:
+    ...
+
+
+@overload
+def align_frames(
+    *frames: pli.LazyFrame,
+    on: str | pli.Expr | Sequence[str] | Sequence[pli.Expr] | Sequence[str | pli.Expr],
+    select: str | pli.Expr | Sequence[str | pli.Expr] | None = None,
+) -> list[pli.LazyFrame]:
+    ...
+
+
+def align_frames(
+    *frames: pli.DataFrame | pli.LazyFrame,
+    on: str | pli.Expr | Sequence[str] | Sequence[pli.Expr] | Sequence[str | pli.Expr],
+    select: str | pli.Expr | Sequence[str | pli.Expr] | None = None,
+) -> list[pli.DataFrame] | list[pli.LazyFrame]:
+    """
+    Align a sequence of frames using the values from one or more columns as a key.
+
+    Frames that do not contain the given key values have rows injected (with nulls
+    filling the non-key columns), and each resulting frame is sorted by the key.
+
+    The original column order of input frames is not changed unless `select` is
+    specified (in which case the final column order is determined from that).
+
+    Note that this does result in a joined frame - you receive the same number
+    of frames back that you passed in, but each is now aligned by key and has
+    the same number of rows.
+
+    Parameters
+    ----------
+    frames
+        sequence of DataFrames or LazyFrames.
+    on
+        one or more columns whose values will be used to align the frames.
+    select
+        optional post-alignment column select to constrain and/or order
+        the columns returned from the newly aligned frames.
+
+    Examples
+    --------
+    >>> # align three frames on the values in column "x", keeping cols "y" and "z"
+    >>> df1, df2, df3 = pl.align_frames(
+    ...     pl.DataFrame({"x": [1, 2, 3], "y": [3.5, 4.0, 1.0], "z": [10.0, 2.5, 1.5]}),
+    ...     pl.DataFrame({"x": [3, 2, 1], "y": [1.0, 8.0, 3.5], "z": [12.0, 1.5, 5.0]}),
+    ...     pl.DataFrame({"x": [2, 3], "y": [5.0, 2.0], "z": [2.0, 2.5]}),
+    ...     on="x",
+    ...     select=["y", "z"],
+    ... )
+    >>> # then use that to calculate the row-wise dot product
+    >>> (df1 * df2 * df3).fill_null(0).select(pl.sum(pl.col("*")).alias("dot"))
+    shape: (3, 1)
+    ┌───────┐
+    │ dot   │
+    │ ---   │
+    │ f64   │
+    ╞═══════╡
+    │ 0.0   │
+    ├╌╌╌╌╌╌╌┤
+    │ 167.5 │
+    ├╌╌╌╌╌╌╌┤
+    │ 47.0  │
+    └───────┘
+
+    """
+    if not frames:
+        return []  # type: ignore[return-value]
+
+    alignment_frame = concat([df.lazy().select(on) for df in frames]).unique(
+        maintain_order=False
+    )
+    aligned_frames = [
+        alignment_frame.join(
+            other=df.lazy(),
+            on=alignment_frame.columns,  # type: ignore[arg-type]
+            how="outer",
+        )
+        .select(df.columns)
+        .sort(by=on)
+        for df in frames
+    ]
+    if select is not None:
+        aligned_frames = [df.select(select) for df in aligned_frames]
+
+    eager = isinstance(frames[0], pli.DataFrame)
```

Comment:
This raises a question: what happens if you pass in a mix of `LazyFrame` and `DataFrame`? According to the typing, this is not officially supported but we should make sure an error is raised -- otherwise users will come to depend on a certain behavior.

I think we could go so far as to traverse the entire list and verify they are all a single type (raise an error if not).

Another option could be to traverse the whole list, and if at least one `DataFrame` is detected the result is eager (I like this one less).

Reply:
Completely agree; passing in a sequence of mixed-type frames feels like an error.

Reply:
Updated everything accordingly